### ISSUE TITLE
Hang in Hercules taking a Standalone Dump

### ISSUE THREAD

**Issue by davekreiss** [SDL-Hercules-390/hyperion#277]
David,

I am running MVS 3.8 with modifications and took a wait state 009064 (S0C4 in NIP) during IPL to a 3390 SYSRES. I have previously built a standalone dump on a 3350 (device 153) and written 4 tape marks to the output dump aws tape (a 3420 at 480). After the wait state I do a store status (STORE command) successfully. I then mount the standalone dump output tape on 480 (via a devinit). I then IPL 153 and Hercules hangs. I have done this quite a few times and did one time get a dump on 480. Don't ask what I did correctly or why it worked that one time, I don't know what I might have done differently.

So again here is what I did:

IPL 19F got disabled wait state 00020000 80090064
store
devinit 480 "\Program Files\Hercules\Tape\sad.aws" awstape
IPL 153
I get a prompt on the 3277 at 010 to which I reply
Instruction counter does nothing and Hercules console doesn't respond. I entered several quit commands to no avail.
I do a windows close (right click on ICON) and select close.
Enclosed is the Hercules log, my IPL batch file, the configuration in a zip file.

I am running the latest (I think) pre-built Hercules-4.2.1.9826 in a DOS window.
Operating system is Windows 10 latest (build 1909).

Let me know what else you might need.

Dave Kreiss
[Hercules.Hang.Doc.zip](https://github.com/SDL-Hercules-390/hyperion/files/4137844/Hercules.Hang.Doc.zip)

**Comment by Fish-Git** [comment]
I would like to try and recreate this hang:

1. Please upload your 153 dasd and your 480 awstape to my ftp website's "incoming" folder:

   ftp://www.softdevlabs.com
   ftp://www.softdevlabs.com/incoming/

> I get a prompt on the 3277 at 010 to which I reply

2. What is the prompt and what was your reply?

3. I noticed the paths in both your batch file as well as in your `devinit` command does not specify a drive letter. Have you tried specifying a drive letter? Is the behavior any different? (e.g. in your `IPLMVSBLD.bat` script, instead of specifying:

```
"\Program Files\Hercules\Hercules\hercules" -f "\Program Files\Hercules\Conf\MVSBLD.conf" >hercules.log
```

Specify a drive letter instead, e.g.:

<pre>
"C:\Program Files\Hercules\Hercules\hercules.exe" -f "C:\Program Files\Hercules\Conf\MVSBLD.conf" > "C:\Program Files\Hercules\<b>Logs</b>\hercules.log"
</pre>

And in your `devinit` command:

```
devinit 480 "C:\Program Files\Hercules\Tape\sad.aws" awstape
```

Is the behavior still the same? Does it still hang? Or does it work any better?

**Comment by davekreiss** [comment]
I thought I'd try to create a much smaller scenario by just including the  MVS 3.8 IPL volume (which gets the wait state) and the SAD disk volume. So I created that config and went through the IPL of 19F encountered wait state,  did store, mounted SAD dump tape and IPLed 153. The standalone dump went flawlessly.  This suggests the hang is related to the Hercules configuration in some way.

So I went back to the original configuration for my MVS 3.8 system, followed the same procedure above and after prompt for the output dump device encountered the originally reported hang in the Hercules console.

I took a dump of the three tasks associated with Hercules. Also took screen prints of Task Manager, Hercules console and tn3270 session.  Zipped together the following and loaded to your incoming ftp site. 
cmd.DMP
conhost.DMP
Console.DMP 
Hercules Console.jpg
Console.jpg
Task Manager.jpg
hercules.log
IPLMVSBLD.bat
SADUMP.153
SADUMP.AWS
MVSBLD.conf

Let me know if you need more info.

Dave

**Comment by Fish-Git** [comment]
> Let me know if you need more info.

Yes: see my previous comment.

**Comment by davekreiss** [comment]
Sorry, not sure what the missing info is.  The adding of drive to commands made no difference.  But I have created a reproduction that consistently fails that only involves two dasd volumes and the tape device.  I've created a zip file with the necessary items and a step by step instructions.  See "SAD Hang Recreate.zip" in your incoming ftp directory and "Hang Process.txt" in that zip file.

Dave

**Comment by Fish-Git** [comment]
> Sorry, not sure what the missing info is.

The information that is missing is the answers to the questions that I asked.

Sentences which end with question marks are requests for specific information. Your subsequent reply/comment did not provide an answer to any of them. Thus yout comment/reply failed to provide the "more information" that I requested. &nbsp; `:)`

> The adding of drive to commands made no difference.

THERE YOU GO! That's what I want to see! An answer to one of my questions! &nbsp; `:)`

That is exactly the type of "more information" that I was wanting that you originally failed to provide in your original reply. You _should_ have provided that information (provided that answer) in your original reply, but you failed to do so.

I don't ask questions for fun, Dave. The questions I ask are not rhetorical. They are virtually always requests for very specific information. You need to pay attention to the questions I and others ask and try to answer them. Doing so is the only way progress can be made when one is trying to resolve a problem. It's to your own benefit after all. You're the one that's asking for help. If you want help with your problem you need to try and provide the information that others ask for, which means answering their questions.

End of lesson. The point is now moot. You HAVE (finally!) answered one of my questions: adding the drive letter to your paths did not make any difference. It did not change the behavior. Thank you. That's good to know. That's one more unknown out of the way.

As far as my other questions are concerned, it appears you have indeed answered those as well ... _in your .zip file that you uploaded to my ftp site._

I expected them to be answered in your github comment reply, but no nevermind. You _did_ answer them. Thank you.

I will try to recreate your problem (the Hercules hang) and will report back here if/when I discover anything. Fair enough?

And my apologies for lecturing you, but you know, I'm human. Sometimes such things "get real old real fast", you know? It happens a lot with some people. Not necessarily you, but it has happened far too often with others. You ask them questions in your attempt to try and help them and it's like they completely ignore you. It's like pulling teeth to get any helpful information out of them. Yes I know the question I asked you in this case was rather trivial and not such a big deal (getting the answer wasn't really _that_ important in this particular case), but it's the _point_ of the matter! You know? When questions are asked one usually expects an answer!

Well, there I go rambling and lecturing again... &nbsp;Sorry!

**Comment by Fish-Git** [comment]
@davekreiss

Dave,

FYI: I _am_ able to reproduce your hang and am working on a fix but it might take me a while yet due to current workload.

**Comment by Fish-Git** [comment]
> I _am_ able to reproduce your hang and am working on a fix...

And by that I mean I _have_ indeed found the bug (i.e. I do know what's causing the problem), but I'm not quite sure yet how to go about fixing it, so it might take me some head-scratching not to mention some possible coordination with a fellow Hercules developer whose coding change to fix a problem he was having is what is now causing _your_ problem.

**Bottom line:** fixing the issue might take a while. &nbsp; `:(`

How long? I don't know.

But the good news is, as I said, I _am_ able to easily recreate your hang and I _do_ know exactly what's causing it, so it's just a matter of time before it gets fixed (which as we all know always seems to be in high demand and short supply).

**Comment by davekreiss** [comment]
I am glad  you have a handle on the cause.  Thanks for the news.  

Dave

**Comment by Fish-Git** [comment]
Dave, (@davekreiss)

Peter's recent `channel.c` commit (8ac28edf9d77bfe48d124443f564562b17f8459f) has now fixed the bug that was causing this problem of yours.

If you are able to build Hercules for yourself, do a git pull to pickup the changes and then rebuild. Otherwise you'll just have to wait until 4.3 is eventually released.

Closing issue.

**Comment by davekreiss** [comment]
Thanks so much for the solution.  I haven't built Hercules, beyond my knowledge so I'll wait for 4.3.  I found a work around by placing Hercules in step mode during the prompts and so was able to take my stand alone dumps.
Dave

**Comment by ivan-w** [comment]
One of the thing I ran into is that there is a mixture of the use of "TCPNJE" as if it where a "typedef" and "struct TCPNJE". Possibly some compilers self correct this, but gcc doesn't. Once I fixed that, there was only a couple of compilation errors that were easy to fix (mostly the missing definition of the variable "filename" which is implicitly used in a convenience macro.

So now it compiles, but I haven't tried using it yet.
